src/Engine/Component/ParticleComponents/ParticleEffect.cpp

```diff
@@ -0,0 +1,393 @@
+    }
+    
+
+    initPosition(p);
+
+    initVelocity(p);
+
+    return p;
+}
+
+void ParticleEffect::initParticle(Particle *p, int i, int meshID, int modelTextureID) {
+    p->i = i;
+    p->meshID = meshID;
+    p->modelTextureID = modelTextureID;
+    p->life = 0.0f;
+}
+
+void ParticleEffect::initVelocity(Particle *p) {
+    switch (m_effectParams->type) {
+        case ParticleEffect::SPHERE : {
+            sphereMotion(p);
+            break;
+        }
+        case ParticleEffect::DISK : {
+            diskMotion(p);
+            break;
+        }
+        case ParticleEffect::CONE: {
+            coneMotion(p);
+            break;
+        }
+    }
+
+    //Add random variance if specified
+    if (m_effectParams->variance != 0.0f) {
+        glm::vec3 randVec = m_effectParams->variance * normalize(glm::ballRand(1.0f));
+        p->velocity += randVec;
+    }
+
+    if (m_effectParams->randomDistribution) {
+        p->velocity *= getRandom(0.25f, 1.0f);
+    }
+
+    //set speed
+    p->velocity *= m_effectParams->magnitude;
+}
+
+void ParticleEffect::initPosition(Particle *p) {
+    p->position = m_anchor + m_offset + m_life * m_velocity;
+}
+
+Vector<glm::vec3> * ParticleEffect::getActiveParticlePositions() {
+    Vector<glm::vec3> *positions = new Vector<glm::vec3>();
+    for (int i = 0; i < m_effectParams->n; i++){
+        if (m_particles[i]->active) {
+            positions->push_back(m_particles[i]->position);
+        }
+    }
+    return positions;
+}
+
+Vector<int> * ParticleEffect::getActiveParticleOrientationIDs() {
+    Vector<int> *orientationIDs = new Vector<int>();
+    for (int i = 0; i < m_effectParams->n; i++) {
+        if (m_particles[i]->active) {
+            orientationIDs->push_back(m_particles[i]->orientationID);
+        }
+    }
+    return orientationIDs;
+}
+
+Vector<int> ParticleEffect::getActiveMap() {
+    Vector<int> map = Vector<int>();
+    for (int i = 0; i < m_effectParams->n; i++) {
+        if (m_particles[i]->active) {
+            map.push_back(i);
+        }
+    }
+    return map;
+}
+
+float ParticleEffect::getNextActivation() {
+    if (m_effectParams->rate == 0.0f) {
+        return 0.0f;
+    }
+    else {
+        return 1.0f / m_effectParams->rate;
+    }
+}
+
+Mesh* ParticleEffect::getMesh(int i) {
+    return m_effectParams->meshes->at(i);
+
+}
+
+ModelTexture* ParticleEffect::getModelTexture(int i) {
+    return m_effectParams->textures->at(i);
+}
+
+ParticleEffect::EffectParams* ParticleEffect::createEffectParams(
+    ParticleEffect::Type type,
+    int n,
+    float effectDuration,
+    float particleDuration,
+    int orientations,
+    bool randomDistribution,
+    float variance,
+    float rate,
+    float angle,
+    bool loop,
+    float magnitude,
+    float attenuation,
+    Vector<glm::vec3>* accelerators,
+    Vector<Mesh *>* meshes,
+    Vector<ModelTexture *>* textures
+){
+    ParticleEffect::EffectParams *effectParams = new ParticleEffect::EffectParams();
```

Comment:
Is this being deleted anywhere?

Reply:
No. Are you getting at that I should use a reference? Because those are a headache to work with.

Reply:
I'm saying if the life span of a ParticleEffect object is shorter than the life span of the entire program, then they be deleted at some point.

Reply:
see million other comments about vectors